### PR TITLE
store-gateway: Add services for BucketStore(s)

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
#### What this PR does

This PR is the first in the refactoring for https://github.com/grafana/mimir/issues/8389. In this PR

1. `BucketStore` (single tenant) and `BucketStores` (multi-tenant) are now `services.Service`, which means they can be properly started and stopped
  2. `BucketStore` currently does nothing on its lifecycle. This is because none of its dependencies are services.
  3. `BucketStores` performs its initial sync in its startup routine and shuts down all of the `BucketStore` instances it hosts in its shutdown routine.
4. Add lifecycle handling to various tests
5. Break up `syncUsersBlocksWithRetries` into `ownedUsers` and `syncUsersBlocks`  so that  `TestBucketStores_ownedUsers` only tests what it needs to - user sharding - and not the whole startup procedure.

#### Which issue(s) this PR fixes or relates to

Related to https://github.com/grafana/mimir/issues/8389

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/storegateway/gateway.go [hunk 1]
```diff
@@ -253,7 +253,7 @@ func (g *StoreGateway) starting(ctx context.Context) (err error) {
 	// At this point, if sharding is enabled, the instance is registered with some tokens
 	// and we can run the initial synchronization.
 	g.bucketSync.WithLabelValues(syncReasonInitial).Inc()
-	if err = g.stores.InitialSync(ctx); err != nil {
+	if err = services.StartAndAwaitRunning(ctx, g.stores); err != nil {
 		return errors.Wrap(err, "initial blocks synchronization")
```
Comment (narqo) [inline pkg/storegateway/gateway.go:257]: (nit) I'm not sure if we should aim fixing it in this PR, but the metrics and the comments around this one now looks off. I.e. this code doesn't "know" anything about "initial sync" any more. Maybe a `TODO` comment about the need to refactor it further later would make this chunk more clear.

File: pkg/storegateway/bucket_stores.go [hunk 2]
```diff
@@ -530,6 +548,9 @@ func (u *BucketStores) getOrCreateStore(userID string) (*BucketStore, error) {
 	if err != nil {
 		return nil, err
 	}
+	if err = services.StartAndAwaitRunning(ctx, bs); err != nil {
+		return nil, fmt.Errorf("starting bucket store for tenant %s: %w", userID, err)
+	}
```
[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/storegateway/bucket_stores.go:553]: inside `closeBucketStore` we call `RemoveBlocksAndClose` which in turn stops `bs`

https://github.com/grafana/mimir/blob/aa3b36a026ceb6542c94fc05203fdb6de15b87ff/pkg/storegateway/bucket_stores.go#L437

https://github.com/grafana/mimir/blob/73160bd2bf5718701cb40fd9d6083c82ae8b97f6/pkg/storegateway/bucket.go#L290-L302

File: pkg/storegateway/gateway.go (hunk 1 above)
Comment (dimitarvdimitrov) [reply pkg/storegateway/gateway.go:257]: i think it's in scope of this PR because they weren't outdated before this PR. They're updated now

**Review by narqo — APPROVED** [review]
Works for me 🔥

**Pull request merged**